opds2_import.py

```diff
@@ -0,0 +1,830 @@
+        :type identifier_mapping: Dict
+
+        :param metadata_client: A SimplifiedOPDSLookup object that is used to fill in missing metadata
+        :type metadata_client: SimplifiedOPDSLookup
+
+        :param content_modifier: A function that may modify-in-place representations (such as images and EPUB documents)
+            as they come in from the network.
+        :type content_modifier: Callable
+
+        :param map_from_collection: Identifier mapping
+        :type map_from_collection: Dict
+
+        :param mirrors: A dictionary of different MirrorUploader objects for different purposes
+        :type mirrors: List[MirrorUploader]
+        """
+        super(OPDS2Importer, self).__init__(
+            db,
+            collection,
+            data_source_name,
+            identifier_mapping,
+            http_get,
+            metadata_client,
+            content_modifier,
+            map_from_collection,
+            mirrors,
+        )
+
+        self._logger = logging.getLogger(__name__)
+
+    def _extract_subjects(self, subject):
+        """Extract a list of SubjectData objects from the webpub-manifest-parser's subject.
+
+        :param subject: Parsed subject object
+        :type subject: Union[
+            str,
+            List[Union[str, core_ast.Subject]],
+            core_ast.Subject
+        ]
+
+        :return: List of subjects metadata
+        :rtype: List[SubjectMetadata]
+        """
+        self._logger.debug(
+            u"Started extracting subject metadata from {0}".format(encode(subject))
+        )
+
+        subject_metadata_list = []
+
+        if is_string(subject):
+            subject_metadata = SubjectData(
+                type=None, identifier=None, name=subject, weight=1
+            )
+            subject_metadata_list.append(subject_metadata)
+        elif isinstance(subject, core_ast.Subject):
+            scheme = subject.scheme
+
+            subject_type = Subject.by_uri.get(scheme)
+            if not subject_type:
+                # We can't represent this subject because we don't
+                # know its scheme. Just treat it as a tag.
+                subject_type = Subject.TAG
+
+            subject_metadata = SubjectData(
+                type=subject_type, identifier=subject.code, name=subject.name, weight=1
+            )
+
+            subject_metadata_list.append(subject_metadata)
+        elif isinstance(subject, list):
+            for inner_subject in subject:
+                inner_subject_metadata_list = self._extract_subjects(inner_subject)
+
+                subject_metadata_list.extend(inner_subject_metadata_list)
+
+        self._logger.debug(
+            u"Finished extracting contributor metadata from {0}: {1}".format(
+                encode(subject), encode(subject_metadata_list)
+            )
+        )
+
+        return subject_metadata_list
+
+    def _extract_contributors(self, contributor):
```

Comment:
I think this should take a default role.

Reply:
Fixed it